### PR TITLE
ConsumerGroups Implementation

### PR DESCRIPTION
**Pull request by bretrbs**

### CODE REVIEW

File: ingestors/kafka/src/main/scala/hydra/kafka/endpoints/ConsumerGroupsEndpoint.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            }
+          } ~ pathPrefix("getByTopic" / Segment) { topic =>
+            pathEndOrSingleSlash {
+              Subject.createValidated(topic) match {
+                case Some(subject) =>
+                  onComplete(
+                    Futurable[F].unsafeToFuture(consumerGroupsAlgebra.getConsumersForTopic(subject.value))
+                  ) {
+                    case Success(topicConsumers) =>
+                      addPromHttpMetric("", StatusCodes.OK.toString, "/v2/consumer-groups/getByTopic")
```
Comment (w-aldrich) [inline ingestors/kafka/src/main/scala/hydra/kafka/endpoints/ConsumerGroupsEndpoint.scala:39]: We can add the topic to the metrics here as the first parameter.

File: ingestors/kafka/src/main/scala/hydra/kafka/endpoints/ConsumerGroupsEndpoint.scala [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                  addPromHttpMetric(topic, StatusCodes.BadRequest.toString, "/v2/consumer-groups/getByTopic")
+                  complete(StatusCodes.BadRequest, Subject.invalidFormat)
+              }
+            }
+          }
+        } ~ pathPrefix("v2" / "topics" / "getByConsumerGroupName" / Segment) { consumerGroupName =>
+          pathEndOrSingleSlash {
+            onComplete(Futurable[F].unsafeToFuture(consumerGroupsAlgebra.getTopicsForConsumer(consumerGroupName))) {
+              case Success(topics) =>
+                addPromHttpMetric("", StatusCodes.OK.toString, "/v2/topics/getByConsumerGroupName/")
```
Comment (w-aldrich) [inline ingestors/kafka/src/main/scala/hydra/kafka/endpoints/ConsumerGroupsEndpoint.scala:55]: We can add the consumerGroupName here as the first parameter.

File: ingestors/kafka/src/main/scala/hydra/kafka/model/TopicMetadataV2Transport.scala [hunk 3]
```diff
@@ -105,9 +105,9 @@ object TopicMetadataV2Request {
       }
     }
 
-    val invalidFormat = "Invalid Subject. Subject must start with skills, flow, tech, fin, dvs, or an underscore(_). " +
+    val invalidFormat = "Invalid Topic Name. Topic Name must start with skills, flow, tech, fin, dvs, or an underscore(_). " +
       " It may contain only alphanumeric characters, hyphens(-) and periods(.)" +
-      " and must not contain \".-\", \"-.\", \"..\", or \"--\" anywhere within the topic."
+      " and must not contain consecutive special characters anywhere within the topic name."
```
Comment (w-aldrich) [inline ingestors/kafka/src/main/scala/hydra/kafka/model/TopicMetadataV2Transport.scala:110]: Thanks for these fixes.

**Review by w-aldrich — APPROVED** [review]

File: ingestors/kafka/src/main/scala/hydra/kafka/algebras/ConsumerGroupsAlgebra.scala [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+      _ <- if (isFulfilled) {
+        deferred.complete(map)
+      } else {
+        ConcurrentEffect[F].unit
+      }
+    } yield ()
+
+    onStart *> dvsConsumerOffsetStream.flatMap { case ((key, value, _), (partition, offset)) =>
+      fs2.Stream.eval(TopicConsumerOffset.decode[F](key, value).flatMap { case (topicKey, topicValue) =>
+        if (latestPartitionOffset.get(partition).map(_ - 1).contains(offset) && topicValue.isDefined) {
```
[the remark this answers is not in the document]
Comment (bretrbs) [reply ingestors/kafka/src/main/scala/hydra/kafka/algebras/ConsumerGroupsAlgebra.scala:103]: LatestOffsets returns the next index that WILL be filled, not the latest that was actually filled. That requires us to subtract 1 from the Offset returned from that map.

File: ingestors/kafka/src/main/scala/hydra/kafka/algebras/ConsumerGroupsAlgebra.scala [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+  private[algebras] def getOffsetsToSeekTo[F[_]: ConcurrentEffect](
+                                                          cache: Ref[F, PartitionOffsetMap],
+                                                          latestPartitionOffset: PartitionOffsetMap,
+                                                          deferred: Deferred[F, PartitionOffsetMap],
+                                                          dvsConsumerOffsetStream: fs2.Stream[F, (Record, OffsetInfo)]
+                                                        ): F[Unit] = {
+    def onStart = if (latestPartitionOffset.values.forall(_ == 0L)) deferred.complete(Map()) else ConcurrentEffect[F].unit
+    def isComplete: F[Unit] = for {
+      map <- cache.get
+      isFulfilled = map.keys.size == latestPartitionOffset.values.count(_ > 0)
```
[the remark this answers is not in the document]
Comment (bretrbs) [reply ingestors/kafka/src/main/scala/hydra/kafka/algebras/ConsumerGroupsAlgebra.scala:93]: That check is done before the items are added to the cache. Only the latest offset will be added to the cache.

File: ingestors/kafka/src/main/scala/hydra/kafka/algebras/ConsumerGroupsAlgebra.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (w-aldrich) [reply ingestors/kafka/src/main/scala/hydra/kafka/algebras/ConsumerGroupsAlgebra.scala:103]: Some native clients, set the offset to -2 for resetting to earliest and -1 for resetting to latest offset. If someone does that to this topic, will this negatively affect this logic?

File: ingestors/kafka/src/main/scala/hydra/kafka/endpoints/ConsumerGroupsEndpoint.scala (hunk 1 above)
Comment (bretrbs) [reply ingestors/kafka/src/main/scala/hydra/kafka/endpoints/ConsumerGroupsEndpoint.scala:39]: Will do.

File: ingestors/kafka/src/main/scala/hydra/kafka/endpoints/ConsumerGroupsEndpoint.scala (hunk 2 above)
Comment (bretrbs) [reply ingestors/kafka/src/main/scala/hydra/kafka/endpoints/ConsumerGroupsEndpoint.scala:55]: Will do.

File: ingestors/kafka/src/main/scala/hydra/kafka/algebras/ConsumerGroupsAlgebra.scala [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+      }
+      .through(produce(producerSettings))
+      .compile.drain
+  }
+
+  private def getSerializer[F[_]: Sync, A](schemaRegistryClient: SchemaRegistryClient)(isKey: Boolean): Serializer[F, A] =
+    Serializer.delegate[F, A] {
+      val serializer = {
+        val se = new KafkaAvroSerializer(schemaRegistryClient)
+        se.configure(Map(AbstractKafkaAvroSerDeConfig.SCHEMA_REGISTRY_URL_CONFIG -> "").asJava, isKey)
```
[the remark this answers is not in the document]
Comment (bretrbs) [reply ingestors/kafka/src/main/scala/hydra/kafka/algebras/ConsumerGroupsAlgebra.scala:203]: Empty String is the proper config here. We don't use the value, but the key is required in this Config.

**Pull request merged**